File: boj/4779_칸토어집합.js
const filePath = process.platform === 'linux' ? '/dev/stdin' : 'input.txt';
const input = require('fs').readFileSync(filePath).toString().trim().split('\n').map(Number);

/**
 * 첫번째 솔루션 탑다운 방식
 * @param {} input
 * @returns
 */
function solution(input) {
	return input
		.map((i) => {
			const arr = new Array(Math.pow(3, i)).fill('-');
			return splitStr(arr);
		})
		.join('\n');
	function splitStr(str) {
		if (str.length < 3) return str.join('');
		const len = str.length / 3;
		const left = splitStr(str.slice(0, len));
		const mid = Array(len).fill(' ').join('');
		const right = splitStr(str.slice(str.length - len, str.length));

		return left + mid + right;
	}
}

/**
 * 첫번째 솔루션 수정본
 * @param {} input
 * @returns
 */
function solutionUpdate(input) {
	return input
		.map((i) => {
			return split(i);
		})
		.join('\n');
	function split(i) {
		if (i === 0) return '-';
		return (
			split(i - 1) +
			Array(Math.pow(3, i - 1))
				.fill(' ')
				.join('') +
			split(i - 1)
		);
	}
}

console.log(solutionUpdate(input));

/**
 * 2번째 방식
 * 바텀업 방식
 * @param {} input
 * @returns
 */
function solution2(input) {
	const arr = Array(12).fill(' ');
	arr[0] = '-';

	for (let i = 1; i < input.length; i += 1) {
		arr[i] =
			arr[i - 1] +
			Array(Math.pow(3, i - 1))
				.fill(' ')
				.join('') +
			arr[i - 1];
	}

	return input.map((i) => arr[i]).join('\n');
}

// console.log(solution2(input));

File: boj/4779_칸토어집합.py
import sys

sys.stdin = open("./input.txt", 'r')
input_data = list(map(int, sys.stdin.readlines()))


def solution(input_data):
    answer = []
    for i in input_data:
        string = '-' * pow(3, i)
        answer.append(split_str(string))
    return '\n'.join(answer)
# print(solution(input_data))


def solution_update(input_data):
    def func(k):
        if k == 0:
            return '-'
        return func(k-1) + (' ' * (3 ** (k-1))) + func(k-1)

    for i in input_data:
        print(func(i))


print(solution_update(input_data))


def split_str(string):
    split_len = int(len(string) / 3)
    if len(string) < 3:
        return string

    left = split_str(string[0:int(split_len)])
    mid = ' ' * split_len
    right = split_str(string[len(string)-split_len:])

    return left+mid+right


# print(solution(input_data))
